r2-core/src/test/java/test/r2/transport/http/client/TestAsyncPool.java

```diff
@@ -570,6 +580,389 @@ public void testWaitTimeStats() throws Exception
+   *      rate limiter will Queue the creation requests once it reached its maximum concurrency configured.
+   *  C = Ih Phase C, N number of objects are returned to the pool which are created in Phase A, this will make
+   *      the number of idle objects in the pool as N.
+   *  D = In Phase D, All the object creation blocked in Phase B will get un blocked and create number of new objects
+   *      that are equal to the rate limiter concurrency. When rate limiter executes the queued creation requests - it
+   *      should ignore the creation requests as there are no object waiters in the pool and thus effectively only
+   *      creating the absolute minimum required count (N+Concurrency)
+   *
+   * @param numberOfCheckouts the N number of checkout operations that will be performed in phase A & B
+   * @param poolSize the maximum Object Pool Size
+   * @param concurrency the maximum allowed concurrent object creation
+   */
+  @Test(dataProvider = "dataProvider")
+  public void testObjectsAreNotCreatedWhenThereAreNoWaiters(int numberOfCheckouts, int poolSize, int concurrency)
+  {
+    CreationBlockableSynchronousLifecycle blockableObjectCreator =
+        new CreationBlockableSynchronousLifecycle(numberOfCheckouts, concurrency);
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(500);
+    RateLimiter rateLimiter = new ExponentialBackOffRateLimiter(0, 5000,
+        10, executor, concurrency);
+
+    final AsyncPool<Object> pool = new AsyncPoolImpl<Object>("object pool",
+        blockableObjectCreator,
+        poolSize,
+        Integer.MAX_VALUE,
+        _executor,
+        Integer.MAX_VALUE,
+        AsyncPoolImpl.Strategy.MRU,
+        0, rateLimiter
+    );
+
+    pool.start();
+
+    // Phase A : Checking out object 'numberOfCheckout' times !
+    List<Object> checkedOutObjects = performCheckout(numberOfCheckouts, pool);
+
+    // Phase B : Blocking object creation and performing the checkout 'numberOfCheckout' times again
+    blockableObjectCreator.blockCreation();
+    Future<None> future = performUnblockingCheckout(numberOfCheckouts, numberOfCheckouts, pool);
+    blockableObjectCreator.waitUntilAllBlocked();
+
+    // Phase C : Returning the checkedOut objects from Phase A back to the object pool
+    for (Object checkedOutObject : checkedOutObjects)
+    {
+      pool.put(checkedOutObject);
+    }
+
+    // Phase D : All the object creation in phase B gets unblocked now
+    blockableObjectCreator.unblockCreation();
+    try
+    {
+      // Wait for all object creation to be unblocked
+      future.get(5, TimeUnit.SECONDS);
+    }
+    catch (Exception e)
+    {
+      Assert.fail("Unexpected interruption", e);
+    }
+
+    // Making sure the rate limiter pending tasks are submitted to the executor
+    while (rateLimiter.numberOfPendingTasks() > 0)
+    {
+      try
+      {
+        Thread.sleep(1);
+      }
+      catch (Exception e)
```

Comment:
Here no need of catching all exceptions, we can just specify the ones that Thread.sleep returns